File: planik-android/app/src/main/java/pl/planik/domain/service/PlanService.kt
package pl.planik.domain.service

import androidx.paging.PagingData
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emptyFlow
import pl.planik.domain.model.DayEntry
import pl.planik.domain.model.NewDayEntry
import pl.planik.domain.model.NewPlan
import pl.planik.domain.model.Plan
import pl.planik.domain.model.UpdateDayEntry
import pl.planik.domain.source.PlanLocalSource
import pl.planik.domain.source.UserLocalSource
import javax.inject.Inject

class PlanService @Inject constructor(
  private val userLocalSource: UserLocalSource,
  private val planLocalSource: PlanLocalSource,
) {
  suspend fun hasPlan(): Boolean = planLocalSource.hasPlan()

  fun getCurrentPlan(): Flow<Plan?> {
    val currentUserId = userLocalSource.currentUserId() ?: return emptyFlow()
    return planLocalSource.getCurrentPlan(currentUserId)
  }

  suspend fun getPlan(id: Int): Plan? {
    val currentUserId = userLocalSource.currentUserId() ?: return null
    return planLocalSource.getPlan(id, currentUserId)
  }

  fun pagedPlans(pageSize: Int): Flow<PagingData<Plan>> = planLocalSource.pagedPlans(pageSize)

  suspend fun addPlan(newPlan: NewPlan): Plan? {
    val currentUserId = userLocalSource.currentUserId() ?: return null
    val id = planLocalSource.createPlan(currentUserId, newPlan)
    userLocalSource.setHasPlan(true)
    return getPlan(id)
  }

  suspend fun updatePlan(plan: Plan): Plan? {
    val id = planLocalSource.updatePlan(plan) ?: return null
    return getPlan(id)
  }

  suspend fun deletePlan(id: Int): Plan? {
    val plan = getPlan(id)
    planLocalSource.deletePlan(id)
    return plan
  }

  suspend fun getDayEntry(dayEntryId: Int, planId: Int): DayEntry? {
    return planLocalSource.getDayEntry(dayEntryId, planId)
  }

  suspend fun addDayEntry(planId: Int, newDayEntry: NewDayEntry): Int {
    return planLocalSource.addDayEntry(planId, newDayEntry)
  }

  suspend fun updateDayEntry(dayEntryId: Int, planId: Int, dayEntry: UpdateDayEntry): Int? {
    return planLocalSource.updateDayEntry(dayEntryId, planId, dayEntry)
  }
}

File: planik-flutter/planik/lib/domain/service/plan_service.dart
import 'package:fpdart/fpdart.dart';
import 'package:planik/domain/failure/general_failure.dart';
import 'package:planik/domain/model/full_plan.dart';
import 'package:planik/domain/model/plan.dart';
import 'package:planik/domain/source/plan_local_source.dart';
import 'package:planik/domain/source/user_local_source.dart';

class PlanService {
  const PlanService({
    required PlanLocalSource planLocalSource,
    required UserLocalSource userLocalSource,
  })  : _planLocalSource = planLocalSource,
        _userLocalSource = userLocalSource;

  final PlanLocalSource _planLocalSource;
  final UserLocalSource _userLocalSource;

  TaskEither<GeneralFailure, FullPlan> getCurrentPlan() {
    return _userLocalSource.currentPlanId().flatMap((it) => _planLocalSource.getPlan(id: it));
  }

  TaskEither<GeneralFailure, FullPlan> getPlan({required id}) {
    return _planLocalSource.getPlan(id: id);
  }

  TaskEither<GeneralFailure, FullPlan> addPlan({
    String name = "",
    bool current = true,
  }) {
    return _userLocalSource.currentUserId().flatMap(
          (currentUserId) => _planLocalSource
              .createPlan(userId: currentUserId, name: name, current: current)
              .flatMap((id) => _planLocalSource.getPlan(id: id))
              .flatMap(
                (fullPlan) => _userLocalSource.setHasPlan(hasPlan: true).map((_) => fullPlan),
              )
              .flatMap(
                (fullPlan) =>
                    _userLocalSource.setCurrentPlanId(id: fullPlan.plan.id).map((_) => fullPlan),
              ),
        );
  }

  TaskEither<GeneralFailure, FullPlan> updatePlan({
    required int id,
    String? name,
    bool? current,
  }) {
    return _planLocalSource
        .updatePlan(id: id, name: name, current: current)
        .flatMap((id) => _planLocalSource.getPlan(id: id))
        .flatMap(
      (fullPlan) {
        if (current == true) {
          return _userLocalSource.setCurrentPlanId(id: fullPlan.plan.id).map((_) => fullPlan);
        } else {
          return TaskEither.right(fullPlan);
        }
      },
    );
  }

  Future<List<Plan>> getPlans({required int pageSize, required int page}) {
    return _planLocalSource
        .getPlans(pageSize: pageSize, page: page)
        .match((l) => <Plan>[], (r) => r)
        .run();
  }
}
